lib/posix-socket/socket_vnops.c

```diff
@@ -0,0 +1,341 @@
+	 * on the application, however, the socket may only receive data after
+	 * the network partner has received the data, which we cannot send.
+	 *
+	 * Since we do not access the vnode in the following code, we can
+	 * unlock it here.
+	 *
+	 * BUT: This allows a different thread to close the vnode in between,
+	 * which would then release it (refcnt = 0). When this thread returns,
+	 * we have a classic use-after-free. Is does not help to increment the
+	 * reference count here before unlocking and then dereference it after
+	 * locking it again. While this would prevent a use-after-free in this
+	 * function the vnode is released as soon as we dereference it, leading
+	 * to a use-after-free in the caller.
+	 *
+	 * A safe solution is to add a vref in all VFS code (i.e., in the
+	 * outmost call frame that works with the vnode), so that it is safe
+	 * to close the vnode in between as the VFS code holds an own reference
+	 * to the vnode as long as it is working with it.
+	 */
+	vn_unlock(vnode);
+
+	ret = posix_socket_write(sock, buf->uio_iov, buf->uio_iovcnt);
+
+	vn_lock(vnode);
+
+	if (unlikely(ret < 0)) {
+		/* TODO: Add fp when write provides it */
+		PSOCKET_ERR("write on socket failed: %d\n", (int)ret);
+		return (int)-ret;
+	}
+
+	buf->uio_resid -= ret;
+
+	return 0;
+}
+
+static int
+posix_socket_vfscore_read(struct vnode *vnode,
+			  struct vfscore_file *fp __maybe_unused,
+			  struct uio *buf, int ioflag __unused)
+{
+	struct posix_socket_file *sock;
+	ssize_t ret;
+
+	UK_ASSERT(vnode->v_data);
+	UK_ASSERT(vnode->v_type == VSOCK);
+
+	sock = (struct posix_socket_file *)vnode->v_data;
+
+	/* TODO: This is not safe. See write() */
+	vn_unlock(vnode);
+
+	ret = posix_socket_read(sock, buf->uio_iov, buf->uio_iovcnt);
+
+	vn_lock(vnode);
+
+	if (unlikely(ret < 0)) {
+		PSOCKET_ERR("read on socket %d failed: %d\n", fp->fd, (int)ret);
+		return (int)-ret;
+	}
+
+	buf->uio_resid -= ret;
+
+	return 0;
+}
+
+static int
+posix_socket_vfscore_ioctl(struct vnode *vnode,
+			   struct vfscore_file *fp __maybe_unused,
+			   unsigned long request, void *buf)
+{
+	struct posix_socket_file *sock;
+	int ret;
+
+	UK_ASSERT(vnode->v_data);
+	UK_ASSERT(vnode->v_type == VSOCK);
+
+	sock = (struct posix_socket_file *)vnode->v_data;
+
+	ret = posix_socket_ioctl(sock, request, buf);
+	if (unlikely(ret < 0)) {
+		PSOCKET_ERR("ioctl on socket %d failed: %d\n", fp->fd,
+			    (int)ret);
+		ret = -ret;
+	}
+
+	return 0;
+}
+
+static int posix_socket_vfscore_poll(struct vnode *vnode, unsigned int *revents,
+				     struct eventpoll_cb *ecb)
+{
+	struct posix_socket_file *sock;
+	int ret;
+
+	UK_ASSERT(vnode->v_data);
+	UK_ASSERT(vnode->v_type == VSOCK);
+
+	sock = (struct posix_socket_file *)vnode->v_data;
+
+	ret = posix_socket_poll(sock, revents, ecb);
+	if (ret < 0)
+		return -ret;
+	if (unlikely(ret < 0)) {
+		PSOCKET_ERR("poll on socket %d failed: %d\n", fp->fd,
+			    (int)ret);
```

Comment:
`fp` is missing here from the signature. Is it lingering in the error?

Reply:
Yap! That needs to be `sock->vfs_file->fd`